pysingfel/particleCollection.py

```diff
@@ -0,0 +1,246 @@
+
+    def bb_aspect(self):
+        """
+        Returns the aspect ratio X:Y:Z of the bounding box.
+        """
+
+        (xmin, xmax), (ymin, ymax), (zmin, zmax) = self.get_bb()
+        xsize = xmax-xmin
+        ysize = ymax-ymin
+        zsize = zmax-zmin
+
+        return (xsize, ysize, zsize)/min(xsize, ysize, zsize)
+    
+    def max_radius(particles):
+        radius_current = 0
+        for particle in particles:
+            radius_arr = particle.atom_pos - np.mean(particle.atom_pos, axis=0)
+            for row in radius_arr:
+                radius = np.sqrt(row[0]**2+row[1]**2+row[2]**2)
+                if radius > radius_current:
+                    radius_current = radius
+        radius_max = radius_current
+        return radius_max
+
+    def add(self, pos, radius, density=1., check=False):
+        """
+        Add a particle to the particle collection.
+        If check=True the distance between the proposed particle and each other
+        is checked so see if they overlap. If so, False is returned.
+        """
+
+        if check:
+            if not self.check(pos, radius):
+                return False
+
+        if len(pos) != 3:
+            print('ERROR: particle position should be given as an x,y,z tuple')
+            return None
+
+        radius = float(radius)
+
+        self.pos[self.count] = np.array(pos)
+        self.radius[self.count] = radius
+        self.volume[self.count] = (4./3.)*np.pi*radius**3.
+        self.density[self.count] = density
+        self.mass[self.count] = self.volume[self.count]*density
+
+        self.count += 1
+        self.idx[self.count-1] = self.next_idx
+        self.next_idx += 1
+
+        return True
+
+    def intersect(self, position, direction, closest=True):
+        """
+        Wrapper for line_sphere() that detects if the position passed is for a
+        monomer or an aggregates and handles each case.
+        """
+
+        if type(position)==tuple:
+            position = np.array(position)
+
+        if len(position.shape)==2: # position is an array, i.e. an aggregate
+            # loop over each monomer in the passed aggregate and check if it
+            # intersects any of the monomers already in the domain
+
+            max_dist = 10000. # TODO calculate a sensible value here
+            pc_id = None
+            hits = None
+            monomer_pos = None
+
+            for pos in position:
+
+                ids, dist = self.line_sphere(pos, direction, closest=True, ret_dist=True)
+                if dist is not None:
+                    if dist < max_dist:
+                        max_dist = dist
+                        monomer_pos = pos
+                        pc_id = ids # id of the particle collection agg
+
+            if pc_id is not None:
+                hit = monomer_pos + max_dist * direction # position of closest intersect
+                return pc_id, max_dist, hit
+            else:
+                return None, None, None
```

Comment:
this function returns 2 things, right?

Reply:
if ret_dist=False, return 2 things: ids, hits; 
if ret_dist=True, return three things: pc_id, max_dist, hit